src/libraries/System.Private.CoreLib/src/System/Reflection/Emit/TypeBuilderInstantiation.cs

```diff
@@ -253,7 +253,8 @@ public override bool ContainsGenericParameters
 
         [RequiresUnreferencedCode("If some of the generic arguments are annotated (either with DynamicallyAccessedMembersAttribute, or generic constraints), trimming can't validate that the requirements of those annotations are met.")]
         public override Type MakeGenericType(params Type[] inst) { throw new InvalidOperationException(SR.Format(SR.Arg_NotGenericTypeDefinition, this)); }
-        public override bool IsAssignableFrom([NotNullWhen(true)] Type? c) { throw new NotSupportedException(); }
+
+        public override bool IsAssignableFrom([NotNullWhen(true)] Type? c) => this == c;
```

Comment:
This is a public API that is going to return a wrong answer now. Throwing NotSupportedException is better than returning a wrong answer.

We should either fully implement this API (complicated), avoid calling it from the validation, or drop the offending validation.

Reply:
@jkotas Just so we are all on the same page, the issue with the above implementation is that it is too restrictive and will return `false` in cases where a complete implementation would return `true`, correct?

Reply:
> We should either fully implement this API (complicated), avoid calling it from the validation, or drop the offending validation.

Agree, I would like to keep the validation whenever possible as it would help avoid overriding a method incorrectly. 

I assumed full implementation was not necessary for `TypeBuilderInstantiation` instances, but now I see my assumption did not cover all scenario. 

I will try to fully implement the method.

Reply:
Have you made a progress on fully implementing the method? I think it is very non-trivial to implement it. Deleting the validation that depends on it would be much easier.

Reply:
I think I am close to full implementation: 

```cs
public override bool IsAssignableFrom([NotNullWhen(true)] Type? c)
{
    if (this == c)
    {
        return true;
    }

    if (c == null || !c.IsConstructedGenericType)
    {
        return false;
    }

    Type typeDefinition = c.GetGenericTypeDefinition();

    if (typeDefinition.IsAssignableFrom(_genericType) &&
        _typeArguments.Length == c.GenericTypeArguments.Length &&
            CheckArguments(c.GenericTypeArguments)
    {
        return true;
    }

    foreach (Type iface in typeDefinition.GetInterfaces())
    {
        if (iface.IsConstructedGenericType &&
            iface.GetGenericTypeDefinition().IsAssignableFrom(_genericType) &&
            iface.GenericTypeArguments.Length == _typeArguments.Length &&
            CheckArguments(c.GenericTypeArguments))
        {
            return true;
        }
    }

    if (typeDefinition.BaseType != null &&
        typeDefinition.BaseType.IsConstructedGenericType &&
        typeDefinition.BaseType.IsAssignableFrom(_genericType) &&
        typeDefinition.BaseType.GenericTypeArguments.Length == _typeArguments.Length &&
        CheckArguments(c.GenericTypeArguments))
    {
        return true;
    }

    return false;
}

private bool CheckArguments(Type[] genericTypeArguments)
{
    for (int i = 0; i < _typeArguments.Length; i++)
    {
        if (_typeArguments[i] != genericTypeArguments[i])
        {
            return false;
        }
    }

    return true;
}
```
Though trying to test it hitting an issue with other validation for example the methods return type / parameters validation:
https://github.com/dotnet/runtime/blob/1e9c6a82aca4904828636b3638962c05a5f8d9c8/src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs#L393-L401
These checks gonna fail when `TypeBuilder` involved as generic type parameter, wonder if I should try to at least relax the validation (exclude generic parameter types) or remove all validations 🤔

Reply:
> I think it is very non-trivial to implement it. Deleting the validation that depends on it would be much easier.

What you think about relaxing the validation? I mean having a partial validation might be better than nothing?

Reply:
Sounds good to me.